### PR TITLE
run.next

### PR DESCRIPTION
**Pull request by snewcomer**
Here is a first shot at the `next` decorator.  @pzuraq @buschtoens

### CODE REVIEW

File: addon/index.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import { run } from '@ember/runloop';
+
+function decoratorWithReturnValue(fn) {
+  return decorator(function(target, key, desc/*, params*/) {
+    // set as value on descriptor - a function that when called will 
+    // return run.next(myFunc)
+    return function(...args)  {
+      const { value } = desc;
+      return fn(value, ...args);
+    }
```
Comment (snewcomer) [inline addon/index.js]: oooh I forgot about the `context` here.  I wonder if the sol'n is just `() =>`?

**Comment by snewcomer** [comment]
@pzuraq @buschtoens would be curious as to what your thoughts were on this PR!

File: addon/index.js (hunk 1 above)
Comment (pzuraq) [reply addon/index.js]: I think you could just pass `this` directly to the `fn` here:

```js
return fn(this, value, ...args)
```

File: addon/index.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ *   @bind
+ *   foo() {
+ *     // do something
+ *   }
+ * }
+ * ```
+ * @function
+ */
+export const bind = decorator((target, value, desc) => {
+  return run.bind(target, desc.value);
```
Comment (snewcomer) [inline addon/index.js]: This isn't working atm.  `target` is just an object...not the context of this.

Comment (pzuraq) [reply addon/index.js]: I think `bind` will have to use an initializer to set the function up when the class in instantiated. Let's do this in a separate followup PR

**Comment by rwjblue** [comment]
@pzuraq - I share your concern here. I don't fully understand the desired usage scenario here...

**Comment by pzuraq** [comment]
Where I'm having difficulty drawing a line is with decorators like `@scheduleOnce` and `@once`, which are borderline to me. The nature of these functions is that people mess up using them frequently, along with `throttle` and `debounce`, because they don't realize that they must pass in a static reference to a function. A decorator would cleanly solve this issue:

```js
class Foo {
  @scheduleOnce('actions')
  bar() {
    //...
  }

  baz() {
    this.bar('qux', 123); // scheduled once
    this.bar('qux', 123); // not scheduled again
  }
}

// vs

class Foo {
  bar() {
    //...
  }

  baz() {
    run.scheduleOnce(this, this.bar, 'actions', 'qux', 123); // scheduled once
    run.scheduleOnce(this, this.bar, 'actions', 'qux', 123); // not scheduled again
  }
}
```

I'm iffy on this use case too since scheduling is so tied to the runloop, but it seems related to `throttle` and `debounce` and I like the idea of those as decorators. From there, it's hard to draw a line in the sand between `@scheduleOnce` and `@schedule`, `@next`, etc because they fundamentally _are_ very similar, just without one-timed-ness.

I would be ok with drawing the line at just `@bind`, `@throttle`, and `@debounce`, or even just `@bind` as well. My thinking was now would be a good time to experiment, but we also have a lot of other work to do so maybe better to be conservative.

**Comment by pzuraq** [comment]
Alternatively, we could go without `run` decorators and for `bind` recommend using initializers instead:

```js
class Foo {
  bar = bind(() => {
    //...
  })
}
```

**Comment by pzuraq** [comment]
Spoke with @rwjblue on slack about this, we agree they may be useful and we can keep experimenting for the time being, see how things work out. Thanks again for pushing this through @snewcomer!

**Pull request merged**